mediatek: add support for Mercusys MR80X v3

Schneider-Azima opened the pull request:
This commit adds support for Mercusys MR80X(EU) v3 router.

Device specification:
 - SoC: Mediatek MT7981b, Cortex-A53, 64-bit
 - RAM: 512MB
 - Flash: SPI NAND GigaDevice GD5F1GQ5UEYIGY (128 MB)
 - Ethernet: 4x 100/1000 Mbps LAN1,LAN2,LAN3 & WAN
 - Wireless: 2.4GHz (802.11 b/g/n/ax)
 - Wireless: 5GHz (802.11 a/n/ac/ax)
 - LEDs: 1 orange and 1 green status LEDs, 4 green gpio-controlled LEDs on ethernet ports
 - Buttons: 1 (Reset)
 - Bootloader: Main U-Boot - U-Boot 2022.01-rc4. Additionally, both UBI slots contain "seconduboot" (also U-Boot 2022.01-rc4)

Installation (UART):
 - Place OpenWrt initramfs-kernel image on tftp server with IP 192.168.1.2
 - Attach UART, switch on the router and interrupt the boot process by
   pressing 'Ctrl-C'.
 - Set the uboot environment for startup.
   setenv tp_boot_idx 0; setenv bootcmd bootm 0x46000000; saveenv
   If the bootarg is set to boot from ubi1, also change it to ubi0.
 - Load and run OpenWrt initramfs image.
   setenv serverip 192.168.1.2; setenv ipaddr 192.168.1.1; tftpboot initramfs-kernel.bin; bootm
 - Browse IP 192.168.1.1, upload the 'sysupgrade' image and do upgrade.

Recovery:
 - Press Reset button and power on the router.
 - Navigate to U-Boot recovery web server (http://192.168.1.1/) and upload the OEM firmware.

Stock layout:
0x000000000000-0x000000200000 : "boot"
0x000000200000-0x000000300000 : "u-boot-env"
0x000000300000-0x000003500000 : "ubi0"
0x000003500000-0x000006700000 : "ubi1"
0x000006700000-0x000006f00000 : "userconfig"
0x000006f00000-0x000007300000 : "tp_data"

ubi0/ubi1 format:
U-Boot at boot checks that all volumes are in place: 
+-------------------------------+
| Volume Name: uboot   Vol ID: 0|
| Volume Name: kernel  Vol ID: 1|
| Volume Name: rootfs  Vol ID: 2|
+-------------------------------+

MAC addresses:
+---------+-------------------+-----------+
|            | MAC               | Algorithm |
+---------+-------------------+-----------+
| label         | 94:0C:xx:xx:xx:12 | label      |
| WAN        | 94:0C:xx:xx:xx:13 | label+1  |
| LAN          | 94:0C:xx:xx:xx:12 | label     |
| WLAN 2g | 94:0C:xx:xx:xx:11 | label-1   |
| WLAN 5g | 94:0C:xx:xx:xx:10 | label-2   |
+---------+-------------------+-----------+
label MAC address was found in UBI partition "tp_data", file
"default-mac". Stock eeprom is also there.

Signed-off-by: Schneider Azima <Schneider-Azima12@protonmail.com>

Schneider-Azima commented:
This is a new PR of mercusys mr80x v3. Refer to #17169 for earlier discussions and modifications.
I made two changes form the old one.
1. Use stock eeprom instead of compiled prooduct of mt76, as what have done in mr90x #13575  and ax80v1 #17753 .
2. Add web upgrade support.

I kindly invite you to review this new PR and suggest changes. @blogic @hauke @fede843

musashino205 commented on target/linux/mediatek/dts/mt7981b-mercusys-mr80x-v3.dts:
```diff
@@ -0,0 +1,328 @@
+			reg = <0x6700000 0x800000>;
+		};
+
+		factory:partition@6f00000 {
```
Please this unused DT label `factory:`.

musashino205 commented on target/linux/mediatek/dts/mt7981b-mercusys-mr80x-v3.dts:
```diff
@@ -0,0 +1,328 @@
+		spi-tx-buswidth = <4>;
+		spi-rx-buswidth = <4>;
+
+
```
Please add a missing `partitions {}` node with `compatible = "fixed-partitions";` and move the all `partition@*` nodes into that.

musashino205 commented on target/linux/mediatek/dts/mt7981b-mercusys-mr80x-v3.dts:
```diff
@@ -0,0 +1,328 @@
+
+&wifi {
+	status = "okay";
+};
```
Please add an empty line at the end of file.

musashino205 commented on target/linux/mediatek/filogic/base-files/etc/board.d/02_network:
```diff
@@ -113,6 +113,9 @@ mediatek_setup_interfaces()
 	mediatek,mt7988a-rfb)
 		ucidef_set_interfaces_lan_wan "lan0 lan1 lan2 lan3 eth2" eth1
 		;;
+	mercusys,mr80x-v3)
+		ucidef_set_interfaces_lan_wan "lan1 lan2 lan3" eth1
+		;;
```
This is the same as the existing definition of `abt,asr3000`-`zbtlink,zbt-z8103ax`, please merge with them.

musashino205 commented on target/linux/mediatek/dts/mt7981b-mercusys-mr80x-v3.dts:
```diff
@@ -0,0 +1,328 @@
+		switch@0 {
+			compatible = "mediatek,mt7531";
+			reg = <31>;
+			reset-gpios = <&pio 39 0>;
```
Please use `GPIO_ACTIVE_HIGH` instead of `0`.

musashino205 commented on target/linux/mediatek/dts/mt7981b-mercusys-mr80x-v3.dts:
```diff
@@ -0,0 +1,328 @@
+			partner_gmac = <1>;
+		};
+
+		switch@0 {
```
This shoud be `switch@<hex phy addr>`, so please use `switch@1f` instead.

musashino205 commented on target/linux/mediatek/dts/mt7981b-mercusys-mr80x-v3.dts:
```diff
@@ -0,0 +1,328 @@
+	};
+
+	gpio-keys-polled {
+		compatible = "gpio-keys-polled";
```
Please try to use the `"gpio-keys"` instead. Then, `poll-interval` property can be removed.

musashino205 commented on target/linux/mediatek/dts/mt7981b-mercusys-mr80x-v3.dts:
```diff
@@ -0,0 +1,328 @@
+			label = "reset";
+			linux,code = <KEY_RESTART>;
+			gpios = <&pio 1 GPIO_ACTIVE_LOW>;
+			linux,input-type = <1>;
```
Please use the `EV_KEY` macro instead. Or, `EV_KEY` is a default type of `"gpio-keys"` (and `"gpio-keys-polled"`) and this can be removed.

musashino205 commented on target/linux/mediatek/dts/mt7981b-mercusys-mr80x-v3.dts:
```diff
@@ -0,0 +1,328 @@
+		phy5: phy@5 {
+			compatible = "ethernet-phy-ieee802.3-c45";
+			reg = <5>;
+			reset-gpios = <&pio 14 1>;
```
Please use the `GPIO_ACTIVE_LOW` macro instead of `1`.

Schneider-Azima commented:
Done

musashino205 commented on package/boot/uboot-mediatek/Makefile:
```diff
@@ -340,6 +340,18 @@ define U-Boot/mt7981_openwrt_one-nor
 endef
 
+define U-Boot/mt7981_mercusys_mr80x-v3
+  NAME:=mercusys mr80x-v3
```
```suggestion
  NAME:=Mercusys MR80X v3
```

Gilgamesh-lzq commented:
【自动回复】 您的邮件已收到，谢谢。[Auto-response] Your email have been received, thank you.

musashino205 commented on target/linux/mediatek/image/filogic.mk:
```diff
@@ -1137,6 +1137,21 @@ define Device/mediatek_mt7988a-rfb
 TARGET_DEVICES += mediatek_mt7988a-rfb
 
+define Device/mercusys_mr80x-v3
+  DEVICE_VENDOR := mercusys
+  DEVICE_MODEL := mr80x-v3
```
```suggestion
  DEVICE_VENDOR := Mercusys
  DEVICE_MODEL := MR80X
  DEVICE_VARIANT := v3
```

Schneider-Azima commented:
Done

Schneider-Azima committed:
mediatek: add support for Mercusys MR80X v3

This commit adds support for Mercusys MR80X(EU) v3 router.

Device specification:
 - SoC: Mediatek MT7981b, Cortex-A53, 64-bit
 - RAM: 512MB
 - Flash: SPI NAND GigaDevice GD5F1GQ5UEYIGY (128 MB)
 - Ethernet: 4x 100/1000 Mbps LAN1,LAN2,LAN3 & WAN
 - Wireless: 2.4GHz (802.11 b/g/n/ax)
 - Wireless: 5GHz (802.11 a/n/ac/ax)
 - LEDs: 1 orange and 1 green status LEDs, 4 green gpio-controlled LEDs
   on ethernet ports
 - Buttons: 1 (Reset)
 - Bootloader: Main U-Boot - U-Boot 2022.01-rc4. Additionally, both UBI
   slots contain "seconduboot" (also U-Boot 2022.01-rc4)

Installation (UART):
 - Place OpenWrt initramfs-kernel image on tftp server with IP 192.168.1.2
 - Attach UART, switch on the router and interrupt the boot process by
   pressing 'Ctrl-C'.
 - Set the uboot environment for startup.
   setenv tp_boot_idx 0; setenv bootcmd bootm 0x46000000; saveenv
   If the bootarg is set to boot from ubi1, also change it to ubi0.
 - Load and run OpenWrt initramfs image.
   setenv serverip 192.168.1.2; setenv ipaddr 192.168.1.1; tftpboot initramfs-kernel.bin; bootm
 - Browse IP 192.168.1.1, upload the 'sysupgrade' image and do upgrade.

Recovery:
 - Press Reset button and power on the router.
 - Navigate to U-Boot recovery web server (http://192.168.1.1/) and
   upload the OEM firmware.

Stock layout:
0x000000000000-0x000000200000 : "boot"
0x000000200000-0x000000300000 : "u-boot-env"
0x000000300000-0x000003500000 : "ubi0"
0x000003500000-0x000006700000 : "ubi1"
0x000006700000-0x000006f00000 : "userconfig"
0x000006f00000-0x000007300000 : "tp_data"

ubi0/ubi1 format:
U-Boot at boot checks that all volumes are in place:
+-------------------------------+
| Volume Name: uboot   Vol ID: 0|
| Volume Name: kernel  Vol ID: 1|
| Volume Name: rootfs  Vol ID: 2|
+-------------------------------+

MAC addresses:
+---------+-------------------+-----------+
|         | MAC               | Algorithm |
+---------+-------------------+-----------+
| label   | 94:0C:xx:xx:xx:12 | label     |
| WAN     | 94:0C:xx:xx:xx:13 | label+1   |
| LAN     | 94:0C:xx:xx:xx:12 | label     |
| WLAN 2g | 94:0C:xx:xx:xx:11 | label-1   |
| WLAN 5g | 94:0C:xx:xx:xx:10 | label-2   |
+---------+-------------------+-----------+
label MAC address was found in UBI partition "tp_data", file
"default-mac".

Signed-off-by: Schneider Azima <Schneider-Azima12@protonmail.com>
Link: https://github.com/openwrt/openwrt/pull/18181
Signed-off-by: Hauke Mehrtens <hauke@hauke-m.de>

Schneider-Azima commented:
Fixed a spelling mistake in git name

DragonBluep commented on package/boot/uboot-mediatek/patches/445_add_mercusys_mr80xv3_dts_and_target.patch:
```diff
@@ -0,0 +1,326 @@
+--- /dev/null
++++ b/configs/mt7981_mercusys_mr80xv3_defconfig
```
Please add symbol `CONFIG_SERIAL_RX_BUFFER=y`. ref: https://github.com/openwrt/openwrt/commit/3d8d807373746d6b3c1674bc450d78090954346f

Schneider-Azima commented:
Done

csharper2005 commented on target/linux/mediatek/dts/mt7981b-mercusys-mr80x-v3.dts:
```diff
@@ -0,0 +1,331 @@
+		compatible = "mediatek,eth-mac";
+		reg = <1>;
+		phy-mode = "gmii";
+		nick-index = <4>;
```
Do you really need this `nick-index = <4>;`?

csharper2005 reviewed (commented):
And... How do setting up LAN and WAN MACs?

csharper2005 commented on target/linux/mediatek/image/filogic.mk:
```diff
@@ -1137,6 +1137,22 @@ define Device/mediatek_mt7988a-rfb
 TARGET_DEVICES += mediatek_mt7988a-rfb
 
+define Device/mercusys_mr80x-v3
+  DEVICE_VENDOR := Mercusys
```
Please use CAPS. Proof - https://www.mercusys.com

Schneider-Azima commented on target/linux/mediatek/filogic/base-files/lib/upgrade/platform.sh:
```diff
@@ -134,6 +134,12 @@ platform_do_upgrade() {
+		mtd erase ubi0
+		mtd write "$1" ubi0
+		mtd erase ubi1
+		mtd write "$1" ubi1
```
We did not replace the uboot of the stock firmware, it may boot from ubi0 or ubi1. Maybe we need a way to force boot from ubi0. Otherwise, I prefer both be replaced to ensure a successful upgrade.

Schneider-Azima replied:
Thanks to @RolandoMagico , there is indeed a variable in uboot-env that indicates which partition to load, ubi0 or ubi1. So we don't have to write to both slots now.

Schneider-Azima commented on target/linux/mediatek/filogic/base-files/etc/hotplug.d/ieee80211/11_fix_wifi_mac:
```diff
@@ -118,6 +118,13 @@ case "$board" in
 		[ "$PHYNBR" = "1" ] && \
 			macaddr_setbit_la "$(mtd_get_mac_binary rf-eeprom 0x4)" > /sys${DEVPATH}/macaddress
 		;;
+	mercusys,mr80x-v3)
```
We need to replace "-" with ":" to set the MAC correctly. Otherwise, the MAC setting fails and the wifi does not work.

Schneider-Azima commented on target/linux/mediatek/dts/mt7981b-mercusys-mr80x-v3.dts:
```diff
@@ -0,0 +1,331 @@
+		stdout-path = "serial0:115200n8";
+	};
+
+	memory {
+		reg = <0 0x40000000 0 0x10000000>;
+	};
```
Sorry, my fault. 256MB is an incorrect number, but in this case the kernel can use 512MB memory correctly. I guess maybe the kernel got the correct data from the second uboot. 
So maybe the configuration here can really be dropped. I observed that most of the mtk dts do have memory nodes , so I prefer to correct this data and keep the memory node. What do you think?

musashino205 commented on target/linux/mediatek/dts/mt7981b-mercusys-mr80x-v3.dts:
```diff
@@ -0,0 +1,331 @@
+			reset-assert-us = <600>;
+			reset-deassert-us = <20000>;
+			phy-mode = "2500base-x";
+			partner_gmac = <0>;
```
Is this necessary?

Schneider-Azima replied:
I'll drop it.

musashino205 commented on target/linux/mediatek/dts/mt7981b-mercusys-mr80x-v3.dts:
```diff
@@ -0,0 +1,331 @@
+			compatible = "ethernet-phy-ieee802.3-c45";
+			reg = <6>;
+			phy-mode = "2500base-x";
+			partner_gmac = <1>;
```
ditto

Schneider-Azima replied:
I'll drop it.

Schneider-Azima commented:
> And... How do setting up LAN and WAN MACs?

Maybe by LUCI? For now the LAN/WAN MACs are randomly generated.

Schneider-Azima commented:
Pattern issues all fixed.

aiamadeus commented on package/boot/uboot-mediatek/patches/446_add_second_uboot_mkimage_command.patch:
```diff
@@ -0,0 +1,12 @@
+ u-boot.bin: u-boot-nodtb.bin FORCE
+-	$(call if_changed,copy)
++	$(call if_changed,copy) &&\
++	./tools/mkimage -A arm -T standalone  -C none -n "seconduboot"  -e 0x41e00000 -d ./u-boot.bin $(TMPDIR)/second-uboot.bin
```
Why do you need this?

Schneider-Azima replied:
To create a standalone uboot imgae, and place it in ubi volume 0.
May I change to another way, as a makefile function? Just like [047891a](https://github.com/openwrt/openwrt/pull/17753/commits/047891a98a3da96cc01caf3a0f3ad728aa1a3f06#diff-f54843b2d13758fa12550a0a284276695bc0b66919fc7a192c025f2db196b70f). It's not merged, and there would be a merge conflict....

Schneider-Azima replied:
> Why do you need this?

Now we can make a standalone second-uboot by makefile function. This patch is dropped.

aiamadeus commented on target/linux/mediatek/dts/mt7981b-mercusys-mr80x-v3.dts:
```diff
@@ -0,0 +1,329 @@
+			linux,code = <KEY_RESTART>;
+			gpios = <&pio 1 GPIO_ACTIVE_LOW>;
+		};
+	};
```
Keep a blank line

aiamadeus commented on target/linux/mediatek/dts/mt7981b-mercusys-mr80x-v3.dts:
```diff
@@ -0,0 +1,329 @@
+			color = <LED_COLOR_ID_AMBER>;
+			function = LED_FUNCTION_STATUS;
+			gpios = <&pio 4 GPIO_ACTIVE_LOW>;
+		};
```
Keep a blank line, same for other led-x

aiamadeus commented on target/linux/mediatek/dts/mt7981b-mercusys-mr80x-v3.dts:
```diff
@@ -0,0 +1,329 @@
+
+&afe {
+	pinctrl-names = "default";
+	pinctrl-0 = <&pcm_pins>;
```
Usually routers don't need pcm?

Schneider-Azima replied:
You are right.
afe, i2c0, and "silabs,proslic_spi" are all for sound. They are useless here. I'll drop them.

aiamadeus commented on target/linux/mediatek/dts/mt7981b-mercusys-mr80x-v3.dts:
```diff
@@ -0,0 +1,329 @@
+&i2c0 {
+	pinctrl-names = "default";
+	pinctrl-0 = <&i2c_pins>;
+	status = "disabled";
```
This node can be removed.

aiamadeus commented on target/linux/mediatek/dts/mt7981b-mercusys-mr80x-v3.dts:
```diff
@@ -0,0 +1,329 @@
+		reg = <1>;
+		phy-mode = "gmii";
+		phy-handle = <&int_gbe_phy>;
+		phy-link {
```
Please remove this fixed-link.

aiamadeus commented on target/linux/mediatek/dts/mt7981b-mercusys-mr80x-v3.dts:
```diff
@@ -0,0 +1,329 @@
+			reset-gpios = <&pio 14 GPIO_ACTIVE_LOW>;
+			reset-assert-us = <600>;
+			reset-deassert-us = <20000>;
+			phy-mode = "2500base-x";
```
Please remove phy-mode here.

aiamadeus commented on target/linux/mediatek/dts/mt7981b-mercusys-mr80x-v3.dts:
```diff
@@ -0,0 +1,329 @@
+			phy-mode = "2500base-x";
+		};
+
+		phy6: phy@6 {
```
These two 2.5G PHYs don't actually exist?

Schneider-Azima replied:
You are right. They are now dropped. Thanks.

aiamadeus commented on target/linux/mediatek/dts/mt7981b-mercusys-mr80x-v3.dts:
```diff
@@ -0,0 +1,329 @@
+		reg = <0>;
+		spi-max-frequency = <52000000>;
+		spi-tx-buswidth = <4>;
+		spi-rx-buswidth = <4>;
```
```
		spi-tx-bus-width = <4>;
		spi-rx-bus-width = <4>;
```

aiamadeus commented on target/linux/mediatek/dts/mt7981b-mercusys-mr80x-v3.dts:
```diff
@@ -0,0 +1,329 @@
+	status = "okay";
+
+	proslic_spi: proslic_spi@0 {
+		compatible = "silabs,proslic_spi";
```
It looks like this is copied from the SDK dts and doesn't actually exist?

aiamadeus commented on target/linux/mediatek/image/filogic.mk:
```diff
@@ -1137,6 +1137,22 @@ define Device/mediatek_mt7988a-rfb
+  BLOCKSIZE := 128k
+  PAGESIZE := 2048
+  UBINIZE_OPTS := -E 5
+  KERNEL_IN_UBI := 1
```
Can `KERNEL_IN_UBI` be placed above `UBINIZE_xx`?

Schneider-Azima commented:
@RolandoMagico suggests to add a "label-mac-device" entry in aliases section. The machine's MAC is stored in a ubifs instead of being written directly on the flash. I refer to the methods of other contributors, but I don't understand how to configure it in dts.
If anyone knows more, you are welcomed to leave a tip, thank you.

Schneider-Azima commented:
AFAIK, all issues fixed.
May I kindly ask for a review? Thanks for all of your help! @musashino205 @aiamadeus @DragonBluep @csharper2005

aiamadeus commented on target/linux/mediatek/dts/mt7981b-mercusys-mr80x-v3.dts:
```diff
@@ -0,0 +1,292 @@
+&spi1 {
+	pinctrl-names = "default";
+	pinctrl-0 = <&spic_pins>;
+	status = "okay";
```
Actually no chip? This node can be removed.

aiamadeus commented on target/linux/mediatek/dts/mt7981b-mercusys-mr80x-v3.dts:
```diff
@@ -0,0 +1,292 @@
+		};
+	};
+
+	pcm_pins: pcm-pins-g0 {
```
Pinctrls without references should also be removed.

Schneider-Azima replied:
Is "conf-pu/conf-pd" in "spi0_flash_pins" used for usb or tf card? If so, I'll drop it.

Schneider-Azima commented:
Done

csharper2005 commented on target/linux/mediatek/dts/mt7981b-mercusys-mr80x-v3.dts:
```diff
@@ -0,0 +1,229 @@
+		led-2 {
+			color = <LED_COLOR_ID_GREEN>;
+			function = LED_FUNCTION_WAN;
+			gpios = <&pio 9 GPIO_ACTIVE_LOW>;
```
I suggest sort led blocks by their gpio numbers.

csharper2005 replied:
I suggest sort led blocks by their gpio numbers.

csharper2005 commented on target/linux/mediatek/filogic/base-files/lib/upgrade/platform.sh:
```diff
@@ -134,6 +134,10 @@ platform_do_upgrade() {
 		;;
+	mercusys,mr80x-v3)
+		mtd erase ubi0
+		mtd write "$1" ubi0
```
The configuration still won't be saved. Why can't we use standard sysupgrade process and standard sysupgrade image? This operation does not erase second uboot.

And I'm not sure that is really necessary generate custom second u-boot and then replace it on first install. See similar device - https://openwrt.org/toh/mercusys/mr90x_v1#installation_using_serial_console

Schneider-Azima replied:
> The configuration still won't be saved. Why can't we use standard sysupgrade process and standard sysupgrade image? This operation does not erase second uboot.
> 
> And I'm not sure that is really necessary generate custom second u-boot and then replace it on first install. See similar device - https://openwrt.org/toh/mercusys/mr90x_v1#installation_using_serial_console

Yes we can use standard sysupgrade process, and we can escape seconduboot. Now they are changed to standard process. But the new kernel can never be larger than the old one.

So I made a 'factory' image, as a complete UBI, with custom seconduboot inside. If anyone witness an upgrade failure, there is another chance.

csharper2005 replied:
> Yes we can use standard sysupgrade process, and we can escape seconduboot. Now they are changed to standard process. 

Thanks.

> But the new kernel can never be larger than the old one. So I made a 'factory' image, as a complete UBI, with custom seconduboot inside. If anyone witness an upgrade failure, there is another chance.

Old kernel size should not be a big problem. You can first drop (`ubirmvol` command) all ubi volumes from `ubi0` except the `seconduboot` one.

csharper2005 commented on target/linux/mediatek/dts/mt7981b-mercusys-mr80x-v3.dts:
```diff
@@ -0,0 +1,229 @@
+
+			partition@6f00000 {
+				label = "tp_data";
+				reg = <0x6f00000 0x800000>;
```
Does it make sense to make it read only?

csharper2005 commented on target/linux/mediatek/dts/mt7981b-mercusys-mr80x-v3.dts:
```diff
@@ -0,0 +1,229 @@
+
+			partition@6700000 {
+				label = "userconfig";
+				reg = <0x6700000 0x800000>;
```
Does it make sense to make it read only?

csharper2005 commented on target/linux/mediatek/dts/mt7981b-mercusys-mr80x-v3.dts:
```diff
@@ -0,0 +1,229 @@
+
+			partition@3500000 {
+				label = "ubi1";
+				reg = <0x3500000 0x3200000>;
```
Does it make sense to make it read only?

csharper2005 commented on target/linux/mediatek/dts/mt7981b-mercusys-mr80x-v3.dts:
```diff
@@ -0,0 +1,229 @@
+
+			partition@200000 {
+				label = "u-boot-env";
+				reg = <0x0200000 0x0100000>;
```
Would be nice to configure `uboot-env` package for this device. It alows manage uboot environment variables from OpenWrt. https://openwrt.org/docs/techref/bootloader/uboot.config#accessing_u-boot_environment_variables_in_openwrt

csharper2005 commented:
> Maybe by LUCI? For now the LAN/WAN MACs are randomly generated.

You can set MACs using this function https://github.com/openwrt/openwrt/blob/main/target/linux/mediatek/filogic/base-files/etc/board.d/02_network#L145

Schneider-Azima commented:
> You can set MACs using this function https://github.com/openwrt/openwrt/blob/main/target/linux/mediatek/filogic/base-files/etc/board.d/02_network#L145

Done

Schneider-Azima commented:
All done.
Supported uboot-envtool. Made a standard sysupgrade image. Set LAN MAC. Fixed format issues.

csharper2005 commented on target/linux/mediatek/filogic/base-files/etc/board.d/02_network:
```diff
@@ -166,6 +167,11 @@ mediatek_setup_macs()
+	mercusys,mr80x-v3)
+		mac_dirty=$(cat "/tmp/tp_data/default-mac" | sed -n 's/^'"MAC"'://p')
+		label_mac=$(macaddr_canonicalize "$mac_dirty")
+		lan_mac=$label_mac
```
What about wan mac?

csharper2005 commented on package/boot/uboot-mediatek/patches/445_add_mercusys_mr80xv3_dts_and_target.patch:
```diff
@@ -0,0 +1,327 @@
++	#size-cells = <1>;
++	model = "mediatek mt7981";
++	compatible = "mediatek,mt7981", "mediatek,mt7981-rfb";
++	chosen {
```
Please add an empty line before `chosen`.

Schneider-Azima replied:
> Please add an empty line before `chosen`.

The uboot's DTS seems to have many formatting issues that need to be adjusted, with some redundant content. Since MR80X no longer requires to modify uboot, I would like to remove the uboot commit in this PR. Maybe submit a separate PR about uboot in future?

csharper2005 commented on target/linux/mediatek/dts/mt7981b-mercusys-mr80x-v3.dts:
```diff
@@ -0,0 +1,232 @@
+	compatible = "mercusys,mr80x-v3", "mediatek,mt7981";
+
+	aliases {
+		serial0 = &uart0;
```
The device has 2 status leds. Why not use them?

Schneider-Azima replied:
> The device has 2 status leds. Why not use them?

Any default rules for the colors of boot/failsafe/running/upgrade? I checked all the DTS of mediatek, and it seems that developers can set according to personal preferences? Some models use all available colors. Some models support multiple colors but only use one of them.

aiamadeus replied:
Usually it should be consistent with the stock firmware.

Schneider-Azima replied:
> Usually it should be consistent with the stock firmware.

Thanks. Now it's correctly configured.

csharper2005 commented on target/linux/mediatek/dts/mt7981b-mercusys-mr80x-v3.dts:
```diff
@@ -0,0 +1,232 @@
+			gpios = <&pio 5 GPIO_ACTIVE_LOW>;
+		};
+
+		led-2 {
+			color = <LED_COLOR_ID_GREEN>;
+			function = LED_FUNCTION_LAN;
+			function-enumerator = <1>;
+			gpios = <&pio 6 GPIO_ACTIVE_LOW>;
+		};
+
+		led-3 {
+			color = <LED_COLOR_ID_GREEN>;
+			function = LED_FUNCTION_LAN;
+			function-enumerator = <2>;
+			gpios = <&pio 7 GPIO_ACTIVE_LOW>;
+		};
+
+		led-4 {
+			color = <LED_COLOR_ID_GREEN>;
+			function = LED_FUNCTION_WAN;
+			gpios = <&pio 9 GPIO_ACTIVE_LOW>;
+		};
+
+		led-5 {
+			color = <LED_COLOR_ID_GREEN>;
+			function = LED_FUNCTION_LAN;
+			function-enumerator = <3>;
+			gpios = <&pio 10 GPIO_ACTIVE_LOW>;
+		};
+	};
```
Why don't you use these LEDs?

Look at https://github.com/openwrt/openwrt/blob/a947be41b721a7c521d23981adaf06ebd1ef5dfe/target/linux/mediatek/filogic/base-files/etc/board.d/01_leds

Schneider-Azima commented:
Done

csharper2005 commented:
> Done

Thanks. Looks good to me now. One question about install guide:
```
 - Make sure to build the image with kernel config 'CONFIG_BLK_DEV_RAM' and 'CONFIG_RD_XZ'
   enabled. They are disabled by default.
```
Does this mean that the image downloaded from downloads.openwrt.org won't work and user has to compile?

Schneider-Azima commented:
> Does this mean that the image downloaded from downloads.openwrt.org won't work and user has to compile?

Is it compiled by default config?

I suppose these basic macros should in fact have been enabled, otherwise other models would obviously not be able to use initramfs.

I'm just stating that if users choose to compile by themselves, these macros may need to be enabled manually AFAIK.

hozansahin commented:
@Schneider-Azima Hey! 
I am an end user and I will install OpenWrt for the first time. Is there openwrt support for Mercusys MR80X? 
When I review from this PR I understand that there is support. Which and what do we download as sysupgrade image.  What is UART?  Can you elaborate the manual for someone like me?

Schneider-Azima commented:
@hozansahin Thanks for your interest in this work.

> Is there openwrt support for Mercusys MR80X?

This PR is not merged right now. 6 apporves may take some time.

> Which and what do we download as sysupgrade image.

This PR is submitted to the main branch. If someone is willing to cherry pick it to the 2410 branch, you can download the firmware from `downloads.openwrt.org` later. I've tested on 2410 and it works.

> What is UART?

See https://openwrt.org/docs/techref/hardware/port.serial#router_serial_connector_portspoints

> Can you elaborate the manual for someone like me?

I'd like to provide information for the Wiki page. But I may not have the authority to directly modify it.

hauke commented:
> > Does this mean that the image downloaded from downloads.openwrt.org won't work and user has to compile?
> 
> Is it compiled by default config?

What error message do you get when you do not activate these options?
Could you please post the messages you get from the bootloader and the early kernel messages with and without these options.

Schneider-Azima commented:
> > > Does this mean that the image downloaded from downloads.openwrt.org won't work and user has to compile?
> > 
> > 
> > Is it compiled by default config?
> 
> What error message do you get when you do not activate these options? Could you please post the messages you get from the bootloader and the early kernel messages with and without these options.

I found only `CONFIG_RD_XZ` is actually required. initrd is configured to use xz compression by default

[console log with XZ support.txt](https://github.com/user-attachments/files/19677151/console.log.with.XZ.support.txt)
[console log without XZ support.txt](https://github.com/user-attachments/files/19677153/console.log.without.XZ.support.txt)

initramfs just fail to unpack. rootfs of stock firmware is mounted as root.

csharper2005 commented:
> I found only `CONFIG_RD_XZ` is actually required. initrd is configured to use xz compression by default

I found the issue https://github.com/openwrt/openwrt/issues/17194. Probably we should enable `CONFIG_RD_XZ` or 'CONFIG_RD_LZMA' (https://github.com/openwrt/openwrt/issues/17194#issuecomment-2525233767) for the filogic target. But I'm not sure about that. Let's see what other this thread members think.

Schneider-Azima commented:
> > I found only `CONFIG_RD_XZ` is actually required. initrd is configured to use xz compression by default
> 
> I found the issue #17194. Probably we should enable `CONFIG_RD_XZ` or `CONFIG_RD_LZMA` ([#17194 (comment)](https://github.com/openwrt/openwrt/issues/17194#issuecomment-2525233767)) for the filogic target. But I'm not sure about that. Let's see what other this thread members think about that.

#17194 seems to be silent for several months. May I remove the description of this macro in the commit log, and just ignore this issue?

robimarko commented:
Hm, initramfs images already enable the required initrd decompressor.

Schneider-Azima commented:
> Hm, initramfs images already enable the required initrd decompressor.

Done.
Removed the statement of enabling macros in commit log.

hozansahin commented:
@Schneider-Azima Hey!
so now we have official support? 
so where do we download it from?

Schneider-Azima commented:
This works on 2410. Does 2410 still accept new models? If so, may I create a backport PR?

Schneider-Azima commented:
> Hey! so now we have official support? so where do we download it from?

@hozansahin Hi. Now you can download from [firmware-selector](https://firmware-selector.openwrt.org/?version=SNAPSHOT&target=mediatek%2Ffilogic&id=mercusys_mr80x-v3) or [downloads](https://downloads.openwrt.org/snapshots/targets/mediatek/filogic/).

luci package is not included in initramfs-kernel.bin, so you have to use the following command in init image:
```bash
cd /tmp
wget https://downloads.openwrt.org/snapshots/targets/mediatek/filogic/openwrt-mediatek-filogic-mercusys_mr80x-v3-squashfs-sysupgrade.bin
sysupgrade -n ./openwrt-mediatek-filogic-mercusys_mr80x-v3-squashfs-sysupgrade.bin
```

Note that snapshot version is only for developers or testers. For stable use, please consider release version.

hozansahin commented:
Hey! @Schneider-Azima Will there be support for EU v2 versions?

Schneider-Azima commented:
> Hey! Will there be support for EU v2 versions?

@hozansahin  Sorry I don't have this device.

akostadinov commented:
Possible to install without serial?

Schneider-Azima commented:
> Possible to install without serial?

@akostadinov  Unless stock firmware remove signature check, or allow remote login.

Schneider-Azima commented:
> @Schneider-Azima , can it work like MR90X? I don't have the device, I am considering buying one.

It can't. Can't ssh login to stock firmware.

eerison commented:
Hey @Schneider-Azima 

- Is it need to make some soldering or the pins are already there to use uart adapter?
- do you know if this router has good performance creating some vlans? 
- Could you provide some photos (if it's possible) where the pins are located? (A noob user here 😅)

Thank you for  your PR 👏👏

eerison commented:
Just another question

do you know why. It is possible to flash MR70X on web ui, But MR80X no?
https://github.com/openwrt/openwrt/pull/17965

muriluo commented:
Hi there! I am struggling here to build an OpenWRT image for this router. No matter what I do when I compile, the initramfs-kernel.bin works perfectly, but when I flash the sysupgrade.bin I get the "*** Cannot find volume 'uboot' *** volume uboot read error -19 " error message! I tried a few workarounds, but no success:
MT7981> setenv bootcmd 'ubi part ubi0; ubi read 0x46000000 kernel; bootm 0x46000000'
MT7981> setenv bootargs 'console=ttyS0,115200n1 ubi.mtd=ubi0 root=ubi0:rootfs rootfstype=squashfs'
MT7981> saveenv

Any clues?
